### DISCUSSION TITLE
Can I run it without opening Discord? (Node.js)

### DISCUSSION THREAD

**Discussion by bitmap4** [rigwild/discord-self-bot-console#4] [Q&A] [answered]
I want to run a simple script, but without opening the app. 
```js 
let channelId = '' // channel id
for (i = 0; i < 35; i++ ) {
    await api.sendMessage(channelId, 'pls daily')
    let randomDelay = 86400000 + Math.floor(Math.random() * 10 * 60000)
    await delay(randomDelay) // approx 1 day
}
```
Is it possible with this API?
(I am new to programming and haven't learnt much of backend js)

**Comment by rigwild** [comment, marked as answer]
Hey, this project uses the Discord console as it's easily usable and CORS-friendly.

You could achieve similar functionality using Node.js.

---

**EDIT:**

I added this in [a gist](https://gist.github.com/rigwild/6063278cbc9267e7691c084c67176114), just clone it!

```sh
git clone https://gist.github.com/6063278cbc9267e7691c084c67176114.git discord-self-bot-console-nodejs
cd discord-self-bot-console-nodejs
npm i
```

Then, in `index.js` just add your auth token at the bottom and set if the account is a bot or not. Replace the channel id in the script.

Run with

```sh
node index.js
```

---

Initialize a project

```sh
npm init -y
```

Install [node-fetch](https://www.npmjs.com/package/node-fetch)

```sh
npm i node-fetch@cjs
```

Copy the script, insert your Discord token at the bottom.

Remove lines related to functions used in browser (the Discord client is a browser! Using [Electron](https://www.electronjs.org/))

https://github.com/rigwild/discord-self-bot-console/blob/31e687e58ad37932958f132b79c3d7f067700b04/index.js#L87-L100

Add this at the top

```sh
const fetch = require('node-fetch')
```

Done!
Just add your script at the bottom! Wrap it in an async function to use async/await.

```js
...
  var authHeader = 'blablabla'
}
// End of api

// Your script
;(async () => {
let channelId = '' // channel id
for (i = 0; i < 35; i++ ) {
    await api.sendMessage(channelId, 'pls daily')
    let randomDelay = 86400000 + Math.floor(Math.random() * 10 * 60000)
    await delay(randomDelay) // approx 1 day
}
})()
```

Just tested it, it works perfectly ;)

**Reply by bitmap4** [reply]
I tried this but got an error-
```
internal/modules/cjs/loader.js:1080
      throw new ERR_REQUIRE_ESM(filename, parentPath, packageJsonPath);
      ^

Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: <path>/dc-self-bot/node_modules/node-fetch/src/index.js
require() of ES modules is not supported.
require() of <path>/dc-self-bot/node_modules/node-fetch/src/index.js from <path>/dc-self-bot/index.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename <path>/dc-self-bot/node_modules/node-fetch/src/index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from <path>/dc-self-bot/node_modules/node-fetch/package.json.

    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1080:13)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (<path>dc-self-bot/index.js:1:15)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14) {
  code: 'ERR_REQUIRE_ESM'
}
```

**Reply by rigwild** [reply]
It's because node-fetch now ships ESM only. My bad!

Use

```sh
npm i node-fetch@cjs
```

**Reply by bitmap4** [reply]
Works perfectly. Thanks a lot!

**Reply by WoodierCupd** [reply]
I want to use this script on the same pc with 3 different accounts at the same time. But for some reason the other 2 don't work any more. I tried using 3 different index.js, even cloning the project 3 times in different directions. The weird thing is that the other 2 still log stuff like usual and give no erros. Do you have any idea why?

**Comment by soujiokita** [comment]
Still working thru console browser, but not work in nodejs anymore
`{ message: '401: Unauthorized', code: 0 }` why

**Reply by rigwild** [reply]
Hey, just tried and it works.

> I added this in a gist, just clone it!
> 
> ```shell
> git clone https://gist.github.com/6063278cbc9267e7691c084c67176114.git discord-self-bot-console-nodejs
> cd discord-self-bot-console-nodejs
> npm i
> ```
> 
> Then, in `index.js` just add your auth token and replace the channel id in the script.
> 
> Run with
> 
> ```shell
> node index.js
> ```

Did you add your auth token? To get it :

1. Open Chrome devtools on Discord using `Ctrl + shift + i`
1. Go to the network tab and send a message in any channel/DM
1. A new entry should appear, click it then copy the `Authorization` header (in the `Request Headers` section)
1. Paste it in `authHeader` at the end of the script

**Reply by soujiokita** [reply]
hey thanks, that was token from bot portal i thought
is all endpoint working properly? i just trying with `sendMessage`

**Reply by rigwild** [reply]
This was not meant to be used with a bot account. I just tried to be sure and i get the same error as you (HTTP 401).

I got it working by [prefixing the token with `Bot`](https://discord.com/developers/docs/reference#authentication-example-bot-token-authorization-header) and setting an accepted [user agent](https://discord.com/developers/docs/reference#user-agent).

I updated the gist to do it for you :) https://gist.github.com/rigwild/6063278cbc9267e7691c084c67176114

**Note for future readers:** I tried to run it in the Discord client but it doesn't work as the client forces the user agent, the server then rejects the request. Run in node.js!

**Reply by soujiokita** [reply]
beautiful.. thanks, i just never using selfbot so i thought it token from bot

**Reply by alcat88** [reply]
Hi, i have this issue too (401: Unauthorized) but the token and the channelId are both correct (i tested the script before on the console and it works), any suggestion? Thanks!

**Reply by rigwild** [reply]
> Hi, i have this issue too (401: Unauthorized) but the token and the channelId are both correct (i tested the script before on the console and it works), any suggestion? Thanks!

Are are using a bot? And if yes are you using the latest version of the nodejs script? https://gist.github.com/rigwild/6063278cbc9267e7691c084c67176114

**Reply by alcat88** [reply]
Findout the problem, I didn't set to false the variable isBotAccount...
Sorry, never do this stuff after midnight :')
